src/ocean/net/ssl/SslClientConnection.d

```diff
@@ -0,0 +1,586 @@
+    private import ocean.sys.socket.model.ISocket;
+
+    private import ocean.sys.Epoll: epoll_event_t;
+
+    private alias epoll_event_t.Event Event;
+
+    static import ocean.core.ExceptionDefinitions;
+
+    /***************************************************************************
+
+        Exception class thrown on errors.
+
+    ***************************************************************************/
+
+    public static class SslException :
+        ocean.core.ExceptionDefinitions.IOException
+    {
+        import ocean.core.Exception: ReusableExceptionImplementation;
+
+        mixin ReusableExceptionImplementation!() ReusableImpl;
+
+        /*******************************************************************
+
+            Sets the exception instance.
+
+            Params:
+                file_path = path of the file
+                func_name = name of the method that failed
+                msg = message description of the error
+                file = file where exception is thrown
+                line = line where exception is thrown
+
+        *******************************************************************/
+
+        public typeof(this) set ( cstring host_path,
+                istring func_name ,
+                cstring msg,
+                istring file = __FILE__, long line = __LINE__)
+        {
+            this.error_num = error_num;
+            this.func_name = func_name;
+
+            this.ReusableImpl.set(this.func_name, file, line);
+
+            this.ReusableImpl.append(": ");
+            this.ReusableImpl.append(this.func_name);
+
+            this.ReusableImpl.append(msg);
+
+            this.ReusableImpl.append(" on ").append(host_path);
+
+            return this;
+        }
+    }
+
+
+    /***************************************************************************
+
+        A minimal implementation of ISocket
+
+    ***************************************************************************/
+
+    private static class SimpleSocket : ISocket
+    {
+        public this ()
+        {
+            super (addrinfo.sizeof);
+        }
+
+        public override void formatInfo ( ref char[] buf, bool io_error )
+        {
+
+        }
+    }
+
+    /***************************************************************************
+
+        The socket which is used for the connection
+
+    ***************************************************************************/
+
+    private SimpleSocket socket;
+
+
+    /***************************************************************************
+
+        The SSL connection
+
+    ***************************************************************************/
+
+    private SSL * sslHandle;
+
+
+    /***************************************************************************
+
+        SelectClient used for blocking the calling Task
+
+    ***************************************************************************/
+
+    private TaskSelectClient select_client;
```

Comment:
Initial, high-level design question: you're wrapping a socket and a task select client. Another approach would be to accept those in the ctor (or the I/O methods). Pros / cons?

Reply:
I think that this class really needs to own the socket. It is stored inside `sslHandle`, and if the app played around with the socket, all kinds of weird things would happen.

It is true that the socket is only directly required in the `connect` function, where the `sslHandle` is created. But the `select_client` is required everywhere, and the three members are coupled together in a complicated way.